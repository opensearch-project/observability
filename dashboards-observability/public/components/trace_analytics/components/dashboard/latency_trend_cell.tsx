/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { EuiButtonIcon, EuiFlexGroup, EuiFlexItem, EuiPopover, EuiText } from '@elastic/eui';
import _ from 'lodash';
import React, { useState } from 'react';
import { LatencyPlt, LinePlt } from '../common/plots/latency_trend_plt';

export function LatencyTrendCell({
  item,
  traceGroupName,
}: {
  item: {
    popoverData: Plotly.Data[];
    trendData: Plotly.Data[];
  };
  traceGroupName: string;
}) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
  return (
    <EuiFlexGroup gutterSize="s">
      <EuiFlexItem />
      <EuiFlexItem grow={false}>
        <LinePlt data={item.trendData} />
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiPopover
          ownFocus
          anchorPosition="downCenter"
          button={
            <EuiButtonIcon
              aria-label="Open popover"
              onClick={() => setIsPopoverOpen(true)}
              iconType="magnifyWithPlus"
              size="s"
            />
          }
          isOpen={isPopoverOpen}
          closePopover={() => setIsPopoverOpen(false)}
        >
          <EuiFlexGroup>
            <EuiFlexItem>
              <EuiText size="s">24-hour latency trend</EuiText>
              <EuiText size="s">{traceGroupName}</EuiText>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButtonIcon
                aria-label="Close popover"
                iconType="cross"
                color="text"
                onClick={() => setIsPopoverOpen(false)}
              />
            </EuiFlexItem>
          </EuiFlexGroup>
          <LatencyPlt data={item.popoverData} />
        </EuiPopover>
      </EuiFlexItem>
    </EuiFlexGroup>
  );
}
