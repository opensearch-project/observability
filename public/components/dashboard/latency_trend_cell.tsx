/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiButtonIcon, EuiFlexGroup, EuiFlexItem, EuiPopover, EuiText } from '@elastic/eui';
import _ from 'lodash';
import React, { useState } from 'react';
import { LatencyPlt, LinePlt } from '../common/plots/latency_trend_plt';

export function LatencyTrendCell({
  item,
  traceGroupName,
}: {
  item: {
    popoverData: Plotly.Data[];
    trendData: Plotly.Data[];
  };
  traceGroupName: string;
}) {
  const [isPopoverOpen, setIsPopoverOpen] = useState(false);
  return (
    <EuiFlexGroup gutterSize="s">
      <EuiFlexItem />
      <EuiFlexItem grow={false}>
        <LinePlt data={item.trendData} />
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiPopover
          ownFocus
          anchorPosition="downCenter"
          button={
            <EuiButtonIcon
              aria-label="Open popover"
              onClick={() => setIsPopoverOpen(true)}
              iconType="magnifyWithPlus"
              size="s"
            />
          }
          isOpen={isPopoverOpen}
          closePopover={() => setIsPopoverOpen(false)}
        >
          <EuiFlexGroup>
            <EuiFlexItem>
              <EuiText size="s">24-hour latency trend</EuiText>
              <EuiText size="s">{traceGroupName}</EuiText>
            </EuiFlexItem>
            <EuiFlexItem grow={false}>
              <EuiButtonIcon
                aria-label="Close popover"
                iconType="cross"
                color="text"
                onClick={() => setIsPopoverOpen(false)}
              />
            </EuiFlexItem>
          </EuiFlexGroup>
          <LatencyPlt data={item.popoverData} />
        </EuiPopover>
      </EuiFlexItem>
    </EuiFlexGroup>
  );
}
