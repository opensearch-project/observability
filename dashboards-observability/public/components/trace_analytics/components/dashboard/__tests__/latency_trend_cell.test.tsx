/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { LatencyTrendCell } from '../latency_trend_cell';

describe('Latency trend cell component', () => {
  configure({ adapter: new Adapter() });

  it('renders latency trend cell', () => {
    const item = JSON.parse(
      `{"trendData":[{"x":[1605027600000],"y":[154.71],"type":"scatter","mode":"lines","hoverinfo":"none","line":{"color":"#000000","width":1}}],"popoverData":[{"x":[1605027600000],"y":[154.71],"type":"scatter","mode":"lines+markers","hovertemplate":"%{x}<br>Average latency: %{y}<extra></extra>","hoverlabel":{"bgcolor":"#d7c2ff"},"marker":{"color":"#987dcb","size":8},"line":{"color":"#987dcb","size":2}}]}`
    );
    const wrapper = mount(<LatencyTrendCell item={item} traceGroupName="order" />);
    expect(wrapper).toMatchSnapshot();

    wrapper.find('button[aria-label="Open popover"]').simulate('click');
    wrapper.find('button[aria-label="Close popover"]').simulate('click');
  });
});
