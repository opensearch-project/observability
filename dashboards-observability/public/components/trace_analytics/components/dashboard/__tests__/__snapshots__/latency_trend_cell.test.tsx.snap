// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Latency trend cell component renders latency trend cell 1`] = `
<LatencyTrendCell
  item={
    Object {
      "popoverData": Array [
        Object {
          "hoverlabel": Object {
            "bgcolor": "#d7c2ff",
          },
          "hovertemplate": "%{x}<br>Average latency: %{y}<extra></extra>",
          "line": Object {
            "color": "#987dcb",
            "size": 2,
          },
          "marker": Object {
            "color": "#987dcb",
            "size": 8,
          },
          "mode": "lines+markers",
          "type": "scatter",
          "x": Array [
            1605027600000,
          ],
          "y": Array [
            154.71,
          ],
        },
      ],
      "trendData": Array [
        Object {
          "hoverinfo": "none",
          "line": Object {
            "color": "#000000",
            "width": 1,
          },
          "mode": "lines",
          "type": "scatter",
          "x": Array [
            1605027600000,
          ],
          "y": Array [
            154.71,
          ],
        },
      ],
    }
  }
  traceGroupName="order"
>
  <EuiFlexGroup
    gutterSize="s"
  >
    <div
      className="euiFlexGroup euiFlexGroup--gutterSmall euiFlexGroup--directionRow euiFlexGroup--responsive"
    >
      <EuiFlexItem>
        <div
          className="euiFlexItem"
        />
      </EuiFlexItem>
      <EuiFlexItem
        grow={false}
      >
        <div
          className="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <LinePlt
            data={
              Array [
                Object {
                  "hoverinfo": "none",
                  "line": Object {
                    "color": "#000000",
                    "width": 1,
                  },
                  "mode": "lines",
                  "type": "scatter",
                  "x": Array [
                    1605027600000,
                  ],
                  "y": Array [
                    154.71,
                  ],
                },
              ]
            }
          >
            <div>
              -
            </div>
          </LinePlt>
        </div>
      </EuiFlexItem>
      <EuiFlexItem
        grow={false}
      >
        <div
          className="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <EuiPopover
            anchorPosition="downCenter"
            button={
              <EuiButtonIcon
                aria-label="Open popover"
                iconType="magnifyWithPlus"
                onClick={[Function]}
                size="s"
              />
            }
            closePopover={[Function]}
            display="inlineBlock"
            hasArrow={true}
            isOpen={false}
            ownFocus={true}
            panelPaddingSize="m"
          >
            <EuiOutsideClickDetector
              isDisabled={true}
              onOutsideClick={[Function]}
            >
              <div
                className="euiPopover euiPopover--anchorDownCenter"
                onKeyDown={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchStart={[Function]}
              >
                <div
                  className="euiPopover__anchor"
                >
                  <EuiButtonIcon
                    aria-label="Open popover"
                    iconType="magnifyWithPlus"
                    onClick={[Function]}
                    size="s"
                  >
                    <button
                      aria-label="Open popover"
                      className="euiButtonIcon euiButtonIcon--primary"
                      disabled={false}
                      onClick={[Function]}
                      size="s"
                      type="button"
                    >
                      <EuiIcon
                        aria-hidden="true"
                        className="euiButtonIcon__icon"
                        size="m"
                        type="magnifyWithPlus"
                      >
                        <EuiIconEmpty
                          aria-hidden={true}
                          className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                          focusable="false"
                          role="img"
                          style={null}
                        >
                          <svg
                            aria-hidden={true}
                            className="euiIcon euiIcon--medium euiIcon-isLoading euiButtonIcon__icon"
                            focusable="false"
                            height={16}
                            role="img"
                            style={null}
                            viewBox="0 0 16 16"
                            width={16}
                            xmlns="http://www.w3.org/2000/svg"
                          />
                        </EuiIconEmpty>
                      </EuiIcon>
                    </button>
                  </EuiButtonIcon>
                </div>
              </div>
            </EuiOutsideClickDetector>
          </EuiPopover>
        </div>
      </EuiFlexItem>
    </div>
  </EuiFlexGroup>
</LatencyTrendCell>
`;
